### PR TITLE
fix: check the new location entries in the location service

### PR DESCRIPTION
**Pull request by SWlionmint**
I have added conditions a new location has to fullfill before the location can be added to the database. For these please check the story. I have overlooked that the ticket is a bug, so had it tracked as a feature in my commits.

Refs: iris-gateway/iris-backlog#121

### CODE REVIEW

File: iris-location-service/src/main/java/iris/location_service/service/LocationService.java [hunk 1]
```diff
@@ -29,20 +30,47 @@
...
+		return email.matches(regex);
+	}
 
-		// TODO: Define sensible limits for this API
+	static boolean isValidAndNotNullPhoneNumber(String phoneNumber) {
+		if (phoneNumber == null)
+			return false;
+		String regex =
+			"^(\\+\\d{1,3}( )?)?(\\d{3}[ ]?){2}\\d{3}$" + "|^((((\\()?0\\d{4}(\\))?)|(\\+49 (\\()?\\d{4}(\\))?))([/ -])(\\d{6}(-\\d{2})?))$"
+				+ "|^(\\+\\d{1,3}( )?)?((\\(\\d{3}\\))|\\d{3})[- .]?\\d{3}[- .]?\\d{4}$" + "|^(\\+\\d{1,3}( )?)?(\\d{3}[ ]?)(\\d{2}[ ]?){2}\\d{2}$";
```
Comment (tboegner) [inline iris-location-service/src/main/java/iris/location_service/service/LocationService.java:45]: I fear we are too strict with this regex. I think we should write some dedicated unit tests (also for email) and maybe put the functions to a separate validator class.

**Review by tboegner — CHANGES_REQUESTED** [review]
Especially the regex validation should be checked to avoid rejecting valid entries.

File: iris-location-service/src/test/java/iris/location_service/service/LocationServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	LocationInformation getLocationInformation(
+		String id,
+		String providerId,
+		String name,
+		String publicKey,
+		String officialName,
+		String representative,
+		String ownerEmail,
+		String email,
+		String phone) {
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/test/java/iris/location_service/service/LocationServiceTest.java:234]: I had asked myself the same question as I have seen changes in format a few times. I checked at the time. I am using the formatter rules. No idea why there are changes.

File: iris-location-service/src/main/java/iris/location_service/service/LocationService.java [hunk 3]
```diff
@@ -29,20 +30,47 @@
 
 	private final @NotNull DBSearchIndex index;
 
-	public void addLocations(String providerId, List<LocationInformation> locations) {
-		// TODO: Authenticate API Access
+	static boolean isValidAndNotNullEmail(String email) {
+		if (email == null)
+			return false;
+		String regex = "^[\\w-_\\.+]*[\\w-_\\.]\\@([\\w]+\\.)+[\\w]+[\\w]$";
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/service/LocationService.java:36]: You are in a string there. So the actual regex would be: ^[\w-_\.+]*[\w-_\.]\@([\w]+\.)+[\w]+[\w]$
I have tested it with https://regexr.com/ and my tests seems to work fine.

File: iris-location-service/src/main/java/iris/location_service/service/LocationService.java [hunk 4]
```diff
@@ -29,20 +30,47 @@
...
 		var data = locations.stream().map(entry -> {
-			return getLocationFromLocationInformation(providerId, entry);
-		}).collect(Collectors.toList());
+			Location location = getLocationFromLocationInformation(providerId, entry);
+			if (entry.getName() == null
+				|| location.getContactRepresentative() == null
+				|| (!isValidAndNotNullEmail(location.getContactEmail()) && !isValidAndNotNullPhoneNumber(location.getContactPhone()))) {
+				listOfInvalidLocations.add(entry.getId());
+			}
+			return location;
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/service/LocationService.java]: I have replaced this with a check for validity, a for-each loop and a else condition

File: iris-location-service/src/main/java/iris/location_service/service/LocationService.java [hunk 5]
```diff
@@ -29,20 +30,47 @@
...
+		})
+			.filter(
+				entry -> entry.getName() != null
+					&& entry.getContactRepresentative() != null
+					&& (isValidAndNotNullEmail(entry.getContactEmail()) || isValidAndNotNullPhoneNumber(entry.getContactPhone())))
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/service/LocationService.java]: I have replaced this with a check for validity, a for-each loop and a else condition

File: iris-location-service/src/main/java/iris/location_service/service/LocationService.java [hunk 6]
```diff
@@ -29,20 +30,47 @@
...
+			.collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/service/LocationService.java]: I have replaced this with a check for validity, a for-each loop and a else condition

File: iris-location-service/src/main/java/iris/location_service/jsonrpc/LocationRPCImpl.java [hunk 7]
```diff
@@ -30,8 +31,18 @@
...
 	@Override
 	public String postLocationsToSearchIndex(JsonRpcClientDto client, List<LocationInformation> locationList) {
-		locationService.addLocations(client.getName(), locationList);
-		return "OK";
+		List<String> listOfInvalidAddresses = locationService.addLocations(client.getName(), locationList);
+
+		if (!listOfInvalidAddresses.isEmpty()) {
+			String output = "Invalid Locations detected: ";
+			for (Iterator iterator = listOfInvalidAddresses.iterator(); iterator.hasNext();) {
+				output += (String) iterator.next() + ", ";
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/jsonrpc/LocationRPCImpl.java]: Now uses a StringBuilder

File: iris-location-service/src/main/java/iris/location_service/jsonrpc/LocationRPCImpl.java [hunk 8]
```diff
@@ -30,8 +31,18 @@
 
 	@Override
 	public String postLocationsToSearchIndex(JsonRpcClientDto client, List<LocationInformation> locationList) {
-		locationService.addLocations(client.getName(), locationList);
-		return "OK";
+		List<String> listOfInvalidAddresses = locationService.addLocations(client.getName(), locationList);
+
+		if (!listOfInvalidAddresses.isEmpty()) {
+			String output = "Invalid Locations detected: ";
+			for (Iterator iterator = listOfInvalidAddresses.iterator(); iterator.hasNext();) {
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/jsonrpc/LocationRPCImpl.java]: Removed Iterator

File: iris-location-service/src/main/java/iris/location_service/service/LocationService.java (hunk 1 above)
Comment (SWlionmint) [reply iris-location-service/src/main/java/iris/location_service/service/LocationService.java]: I have adjusted the regex based on the tests. I think the regex allows for a lot of different formats already. If we find something we want to add that can be done later.

**Review by tboegner — APPROVED** [review]
I think we are OK for now, as we ensure to have at least one (formally) valid contact possibilty. But we should keep in mind, that it is possible to store invalid data (e.g. an invalid email address if the phone number is valid).

File: iris-location-service/src/test/java/iris/location_service/service/LocationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+		List<LocationInformation> locationList = new ArrayList<LocationInformation>();
+		locationList.add(
+			getLocationInformation(
+				LOCATION_ID_1,
+				"providerId",
+				LOCATION_NAME,
+				"publicKey",
+				"officialName",
+				REPRESENTATIVE_NAME,
+				"ownerEmail",
```
[the remark this answers is not in the document]
Comment (SWlionmint) [reply iris-location-service/src/test/java/iris/location_service/service/LocationServiceTest.java:104]: I agree we can think about this at a later date. I personally think that this should be checked at the UI side. We just want that it is possible to contact the user, not neccessary to make sure that any way given works.

**Pull request merged**